### PR TITLE
[5.3] Fix empty field dates conversion

### LINKED ISSUE #13613: nullable dates field can't be leave unchanged ie.L set to null through a form (simple to fix)
In case there is nullable field which is in the dates fields and you would like to check it's value than your're not able to do it (ie. from a html form you get ''). The reason since every reference to the field eg:
```
empty($this->myfield)
```
call the Model's getAttributeValue and in this function at the line 2636 $this->asDateTime($value) which fail for empty sting $value.
So this should have to change:
```
if (in_array($key, $this->getDates()) && ! is_null($value)) {
      return $this->asDateTime($value);
}
```
to this:
```
if (in_array($key, $this->getDates()) && ! is_null($value) && ($value !== '')) {
      return $this->asDateTime($value);
}
```
otherwise it's not possible to check the value of the field and replace it with null.
Regards

### PR DESCRIPTION
**Pull request by lfarkas**
Closes #13613.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Any reason you sent this to 5.3, and not an earlier version?

**Comment by lfarkas** [comment]
there is not.
i don't know what's the release policy and whether it's a critical bugfix or not. so i let decided by the laravel developers.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -2607,7 +2607,7 @@ public function getAttributeValue($key)
         // If the attribute is listed as a date, we will convert it to a DateTime
         // instance on retrieval, which makes it quite convenient to work with
         // date fields without having to create a mutator for each property.
-        if (in_array($key, $this->getDates()) && ! is_null($value)) {
+        if (in_array($key, $this->getDates()) && ! is_null($value) && ($value !== '')) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: No need for the extra brackets please. ;)

**Comment by taylorotwell** [comment]
Why are you storing an empty string in a date field?

**Comment by lfarkas** [comment]
it's not that simple. actually there are two problem:

1. supporting nullable value in any kind of field like datetime
2. handle response from html form.

for the first there is another issue: https://github.com/laravel/internals/issues/94
for the second (and also for the first) let me show an example:
suppose we like to store flight in a database. each flight has 4 dates (all of them are in $dates):
- estimated departure, actual departure,
- estimated arrival, actual arrival,
any of them can be null. eg:
- we don't know when just sunday we plan a flight (all is null)
- we have plan sunday 11:00-13:35 is the estimated (2 are null)
- we already takeoff but we still not landig (at least one is null)
etc.

for this we can create a form with 4 datetimepicker js but all of them initially empty and if we do not set these fields the post after the submit send these dates as empty field. in this case laravel automatic date mutators will fail without the above modification. ie. you are not even to be able to test them eg: $this->takeoff will gives and error.

how to store them as null is another story of the above issue.

**Comment by syaiful6** [comment]
i think that's not good idea if you store empty string on nullable column. If it were allowed to be null, there would be two ways to have an value on your db: null and "". This gives you multiple values to check against, which is inefficient and messy. I think you should cast that empty string to null before save it.

**Comment by taylorotwell** [comment]
I agree you should either store null or a valid date.

**Pull request closed without merging**